### PR TITLE
Ground work to enable cloud uploads for Koji composes

### PR DESCRIPTION
**Pull request by thozza**
This PR prepares the ground for supporting cloud uploads for Koji composes.

### Main changes

- `OSBuilJob` implementation is now able to handle any number of upload targets per job in a sequence.
- Upload target errors are not set in the `OSBuildJob` results any more, but in the specific target results. The reason is that now there may be more than one target in the job. All target errors are then attached to the job error as `details` and an appropriate job error is set.
- Uploading back to the worker server is now done using an explicit upload target `org.osbuild.worker.server`, instead of setting the `OSBuildJob.ImageName`. `OSBuildJob.ImageName` was removed. No backward compatibility is handled in this case, because this functionality is used only by Weldr API and there it is expected that the server and worker will be of the same version. I considered using the `org.osbuild.local` target, but intentionally didn't do so. The reason is that it contained many useless fields and in order to delete them and use it, I would have to keep backward compatibility layer to support reading stored compose results in the on-premise store. By using a new target, this is not needed. Also a more (IMHO) self-explanatory target name could be used.
- Related to the previous point, the **composes submitted by Weldr API with upload to cloud already use multiple targets in the job** (2 - upload back to worker server and to the cloud).
- Export `Filename` was moved from target-specific Options to the Target itself. In reality, the export filename is specific to the export used for the target and must be defined for any target regardless of its options type. Backward compatibility is handled on the JSON (Un)Marshal level.
- The `Exports` have been moved from the `OSBuildJob` to the Target. In reality, the export is specific to the target. Backward compatibility is handled on the JSON (Un)Marshal level. Each Target can specify only a single export. Only a single common export is still used for all targets. **Technically, the `OSBuildJob` implementation is able to handle different export per target**, however no API implementation does that at this moment.
- The backward compatibility code handling stream-optimized conversion for VMDK is deleted, because all deployments have been already updated and it is not needed for Weldr due to the expectation of server and worker being of the same version. Also deleting it made the rework to support multiple upload targets simpler.

### Miscellaneous changes

- Target names are defined as constants, instead of using string literals everywhere.
- Some unused functions and structure fields were deleted.
- some variable names in the `OSBuildJob` implementation were renamed to better express their purpose.

### What's left (still WIP and not part of this PR)

- Modify Cloud API to allow upload options for Koji composes.
- Extend the `koji.sh` test case to test upload to cloud for Koji composes.

This pull request includes:

- [x] adequate testing for the new functionality or fixed issue
- [ ] adequate documentation informing people about the change such as
  - [ ] submit a PR for the [guides](https://github.com/osbuild/guides) repository if this PR changed any behavior described there: https://www.osbuild.org/guides/

### CODE REVIEW

**Comment by achilleas-k** [comment]
> warning This PR introduces changes in at least one manifest (when comparing PR HEAD [693a3b5](https://github.com/osbuild/osbuild-composer/commit/693a3b5d4da3e5adbc6e0a743dd5d930e8185649) with the main merge-base [3fcb737](https://github.com/osbuild/osbuild-composer/commit/3fcb737243b742fa9143cd9df068044eeff54eb1)). Please review the changes.

Seems something went wrong with one of the repositories for RHEL 8 aarch64 and it affected the packages for the AMI.
```
Failed to add groups file for repository: 29bdead0aef9110bb7ba53940a6a0bf7207adbe61e7cd48a23451bcdb06ba628 - ERROR: Parser error at line:1 col:0
    not well-formed (invalid token)
```

I've restarted the job to make sure it's all fine.

File: internal/cloudapi/v2/v2_koji_test.go [hunk 1]
```diff
@@ -477,7 +477,7 @@ func TestKojiCompose(t *testing.T) {
 				require.NoError(t, err)
 				jobTarget := osbuildJob.Targets[0].Options.(*target.KojiTargetOptions)
 				require.Equal(t, "koji.example.com", jobTarget.Server)
-				require.Equal(t, "test.img", jobTarget.Filename)
+				require.Equal(t, "test.img", osbuildJob.Targets[0].ExportFilename)
```
Comment (thozza) [inline internal/cloudapi/v2/v2_koji_test.go:480]: I just spotted this change, which is a mistake most probably from rebasing. Will fix it...

File: internal/target/target_test.go [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+			assert.NoError(t, err)
+			assert.EqualValues(t, testCase.expectedTarget, &gotTarget)
+		})
+	}
+}
+
+// Test that that ExportFilename set in the Target get added to the options
+// as Filename.
+// This enables old worker to still pick and be able to handle jobs from new composer.
+func TestOSBuildJobExportsCompatibilityMarshal(t *testing.T) {
```
Comment (thozza) [inline internal/target/target_test.go:187]: c&p error: should be renamed to `TestTargetOptionsFilenameCompatibilityMarshal`

**Review by ondrejbudai — DISMISSED** [review]
This is awesome! More praise and minor comments inline.

File: internal/worker/json.go [hunk 3]
```diff
@@ -20,10 +20,25 @@ type OSBuildJob struct {
...
-	Exports            []string         `json:"export_stages,omitempty"`
 	PipelineNames      *PipelineNames   `json:"pipeline_names,omitempty"`
 }
 
+// OsbuildExports returns a slice of osbuild pipeline names, which should be
+// exported as part of running osbuild image build for the job. The pipeline
+// names are gathered from the targets specified in the job.
+func (j OSBuildJob) OsbuildExports() []string {
+	exports := []string{}
+	seenExports := map[string]bool{}
```
[the remark this answers is not in the document]
Comment (thozza) [reply internal/worker/json.go:31]: I know about this and used it originally. But then I read https://itnext.io/set-in-go-map-bool-and-map-struct-performance-comparison-5315b4b107b and decided that this is more readable and the gain from using empty struct is not big enough.

File: cmd/osbuild-worker/jobimpl-osbuild.go [hunk 4]
```diff
@@ -298,7 +298,8 @@ func (impl *OSBuildJobImpl) Run(job worker.Job) error {
 	// copy pipeline info to the result
 	osbuildJobResult.PipelineNames = jobArgs.PipelineNames
 
-	exports := jobArgs.Exports
+	// get exports for all job's targets
+	exports := jobArgs.OsbuildExports()
 	if len(exports) == 0 {
```
[the remark this answers is not in the document]
Comment (thozza) [reply cmd/osbuild-worker/jobimpl-osbuild.go:303]: Do you mean to remove it completely, so not even check if there are any exports defined? The job will eventually fail in that case anyway, just wanted to make sure I understand your suggestion...

**Pull request merged**